server.py

```diff
@@ -15,44 +13,62 @@
 
 # server class
 class Server:
-    def __init__(self, imagedata):
-        self.board = CanvasBoard(imagedata)
-        self.lock = threading.Lock()
+    def __init__(self):
+        self.__boards = {}
 
-    def update_board(self, diffs: list):
-        self.lock.acquire()
-        self.board.update_board(diffs)
-        self.lock.release()
+    def get_ids(self):
+        return self.__boards.keys()
 
+    def add_board(self, room_id, board):
+        self.__boards[room_id] = board
 
-# instantiate server class for board state
-imagedata = {
-    'width': WIDTH,
-    'height': HEIGHT,
-    'data': [0 for i in range(4 * WIDTH * HEIGHT)]
-}
-server = Server(imagedata)
+    def get_board(self, room_id):
+        return self.__boards[room_id]
+
+    def update_board(self, diffs: list, room_id: str):
+        self.__boards[room_id].update_board(diffs)
+server = Server()
 
 @app.route('/')
 def index():
     return app.send_static_file('index.html')
 
+@app.route('/create')
+def create_room(payload):
+    room = payload['room_id']
+    if room in server.get_ids():
+        abort(Response('Room with given ID already exists'))
+    server.add_board(room, CanvasBoard.create_board(WIDTH, HEIGHT))
+    response = jsonify(room_id=room)
+    # just return response with room id
+    # don't worry about frontend
+    return response
+
 @socketio.on('connect', namespace='/canvas')
 def connect_canvas():
-    # broadcast board upon initial connect at /canvas endpoint
-    # turn board into JSON
-    board = CanvasBoardEncoder().encode(server.board)
-    emit('initialize-board', board)
+    print("connected to websocket")
 
 @socketio.on('disconnect')
 def on_disconnect():
     print("disconnected from websocket")
 
 @socketio.on('send-stroke', namespace="/canvas")
-def handle_send_stroke(stroke):
-    server.update_board(stroke['diffs'])
-    # broadcast the new change
-    emit('broadcast-stroke', stroke, broadcast=True)
+def handle_send_stroke(payload):
+    room_id = payload['room_id']
+    if room_id in server.get_ids():
+        server.update_board(payload['diffs'], room_id)
+        emit('broadcast-stroke', payload['diffs'], room=room_id)
+    else:
+        print("Error: no room found")

```

Comment:
We should also return the error to the client.

Reply:
Abort() function used